src/ecdsa/ellipticcurve.py

```diff
@@ -174,20 +175,40 @@ def __init__(self, curve, x, y, z, order=None, generator=False):
             self.__y = y
             self.__z = z
             self.__order = order
+        self.__generator = generator
         self.__precompute = []
-        if generator:
-            assert order
-            i = 1
-            order *= 2
-            doubler = PointJacobi(curve, x, y, z, order)
-            order *= 2
-            self.__precompute.append((doubler.x(), doubler.y()))
-
-            while i < order:
-                i *= 2
-                doubler = doubler.double().scale()
+
+    def _maybe_precompute(self):
+        if self.__generator:
+            # since we lack promotion of read-locks to write-locks, we do a
+            # "acquire-read-lock check, acquire-write-lock plus recheck" cycle ..
+            try:
+                self._scale_lock.reader_acquire()
```

Comment:
it's `_scale_lock` because it was used for `scale()` but now we also use it for precompute... maybe change its name to `_update_lock`?